core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/promisehandlers/PrepareCheckpointHandler.scala

```diff
@@ -0,0 +1,82 @@
+package edu.uci.ics.amber.engine.architecture.worker.promisehandlers
+
+import edu.uci.ics.amber.engine.architecture.worker.{
+  DataProcessorRPCHandlerInitializer,
+  WorkflowWorker
+}
+import edu.uci.ics.amber.engine.architecture.worker.WorkflowWorker.MainThreadDelegate
+import edu.uci.ics.amber.engine.architecture.worker.promisehandlers.PrepareCheckpointHandler.PrepareCheckpoint
+import edu.uci.ics.amber.engine.common.ambermessage.WorkflowFIFOMessage
+import edu.uci.ics.amber.engine.common.{CheckpointState, CheckpointSupport, SerializedState}
+import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand
+
+import java.util.concurrent.CompletableFuture
+import scala.collection.mutable
+
+object PrepareCheckpointHandler {
+  final case class PrepareCheckpoint(estimationOnly: Boolean) extends ControlCommand[Unit]
+}
+
+trait PrepareCheckpointHandler {
+  this: DataProcessorRPCHandlerInitializer =>
+
+  registerHandler { (msg: PrepareCheckpoint, sender) =>
+    logger.info("Start to take checkpoint")
+    if (!msg.estimationOnly) {
+      dp.serializationCall = () => {
+        serializeWorkerState()
+      }
+    } else {
+      logger.info(s"Checkpoint is estimation-only. do nothing.")
+    }
+  }
+
+  private def serializeWorkerState(): Unit = {
+    val chkpt = new CheckpointState()
+    // 1. serialize DP state
+    chkpt.save(SerializedState.DP_STATE_KEY, this.dp)
+    // checkpoint itself should not be serialized, thus we register it after serialization
+    dp.channelMarkerManager.checkpoints(dp.channelMarkerManager.getContext.marker.id) = chkpt
+    logger.info("Serialized DP state")
+    // 2. serialize operator state
+    if (dp.operatorOpened) {
+      dp.operator match {
```

Comment:
there is an implicit assumption that only the opened operator needs serialization. I am not sure if this assumption holds.

Reply:
true. I plan to remove this assumption for now.